File: examples/python/large_file_upload.py
"""
Large File Upload/Download via Signed URLs

For files > 50MB, use signed URLs instead of the files API.
Signed URLs stream directly to/from the worker — no gRPC size limits.

Usage:
    OPENCOMPUTER_API_KEY=your-key python large_file_upload.py
"""

import asyncio
import hashlib
import os
import time

import httpx
from opencomputer import Sandbox


async def main():
    print("╔═══════════════════════════════════════════╗")
    print("║  Large File Upload via Signed URLs         ║")
    print("╚═══════════════════════════════════════════╝\n")

    sb = await Sandbox.create(timeout=300)
    print(f"Sandbox: {sb.sandbox_id}\n")

    try:
        # ── Upload via signed URL ──────────────────────────────────
        size_mb = 100
        print(f"Generating {size_mb}MB of random data...")
        data = os.urandom(size_mb * 1024 * 1024)
        sha = hashlib.sha256(data).hexdigest()[:16]
        print(f"  SHA-256: {sha}...\n")

        # Get a signed upload URL
        print("Getting signed upload URL...")
        upload_url = await sb.upload_url("/workspace/large.bin")
        print(f"  URL: {upload_url[:80]}...\n")

        # Upload directly to the worker (bypasses gRPC)
        print(f"Uploading {size_mb}MB...")
        t0 = time.time()
        async with httpx.AsyncClient(timeout=120) as client:
            resp = await client.put(
                upload_url,
                content=data,
                headers={"Content-Type": "application/octet-stream"},
            )
        elapsed = time.time() - t0
        mbps = size_mb / elapsed

        if resp.status_code < 300:
            print(f"  ✓ Upload: {elapsed:.1f}s ({mbps:.1f} MB/s)\n")
        else:
            print(f"  ✗ Upload failed: {resp.status_code} {resp.text[:200]}\n")
            return

        # Verify file exists and has correct size
        r = await sb.exec.run("stat -c %s /workspace/large.bin")
        file_size = int(r.stdout.strip())
        print(f"  File size on disk: {file_size / 1024 / 1024:.1f}MB")
        if file_size == len(data):
            print(f"  ✓ Size matches\n")
        else:
            print(f"  ✗ Size mismatch: expected {len(data)}, got {file_size}\n")

        # ── Download via signed URL ────────────────────────────────
        print("Getting signed download URL...")
        download_url = await sb.download_url("/workspace/large.bin")
        print(f"  URL: {download_url[:80]}...\n")

        print(f"Downloading {size_mb}MB...")
        t0 = time.time()
        async with httpx.AsyncClient(timeout=120) as client:
            resp = await client.get(download_url)
        downloaded = resp.content
        elapsed = time.time() - t0
        mbps = size_mb / elapsed

        if resp.status_code == 200:
            print(f"  ✓ Download: {elapsed:.1f}s ({mbps:.1f} MB/s)\n")
        else:
            print(f"  ✗ Download failed: {resp.status_code}\n")
            return

        # Verify hash matches
        dl_sha = hashlib.sha256(downloaded).hexdigest()[:16]
        if sha == dl_sha:
            print(f"  ✓ Hash matches: {dl_sha}...\n")
        else:
            print(f"  ✗ Hash mismatch: {sha}... vs {dl_sha}...\n")

        print("Done — all checks passed!")

    finally:
        await sb.kill()
        print(f"\nSandbox {sb.sandbox_id} destroyed.")


asyncio.run(main())

File: examples/typescript/large-file-upload.ts
/**
 * Large File Upload/Download via Signed URLs
 *
 * For files > 50MB, use signed URLs instead of the files API.
 * Signed URLs stream directly to/from the worker — no gRPC size limits.
 *
 * Usage:
 *   OPENCOMPUTER_API_KEY=your-key npx tsx examples/typescript/large-file-upload.ts
 */

import { Sandbox } from "../../sdks/typescript/src/index.js";
import { createHash, randomBytes } from "crypto";
import { Readable } from "stream";

async function main() {
  console.log("╔═══════════════════════════════════════════╗");
  console.log("║  Large File Upload via Signed URLs         ║");
  console.log("╚═══════════════════════════════════════════╝\n");

  const sb = await Sandbox.create({ timeout: 300 });
  console.log(`Sandbox: ${sb.sandboxId}\n`);

  try {
    // ── Upload via signed URL ──────────────────────────────────
    const sizeMB = 100;
    console.log(`Generating ${sizeMB}MB of random data...`);
    const data = randomBytes(sizeMB * 1024 * 1024);
    const hash = createHash("sha256").update(data).digest("hex").slice(0, 16);
    console.log(`  SHA-256: ${hash}...\n`);

    // Get a signed upload URL
    console.log("Getting signed upload URL...");
    const uploadUrl = await sb.uploadUrl("/workspace/large.bin");
    console.log(`  URL: ${uploadUrl.slice(0, 80)}...\n`);

    // Upload directly to the worker (bypasses gRPC)
    console.log(`Uploading ${sizeMB}MB...`);
    const t0 = Date.now();
    const uploadResp = await fetch(uploadUrl, {
      method: "PUT",
      headers: { "Content-Type": "application/octet-stream" },
      body: data,
    });
    const uploadMs = Date.now() - t0;

    if (uploadResp.ok) {
      const mbps = (sizeMB / (uploadMs / 1000)).toFixed(1);
      console.log(`  ✓ Upload: ${uploadMs}ms (${mbps} MB/s)\n`);
    } else {
      console.log(`  ✗ Upload failed: ${uploadResp.status} ${await uploadResp.text()}\n`);
      return;
    }

    // Verify file exists and has correct size
    const check = await sb.exec.run(`stat -c %s /workspace/large.bin`);
    const fileSize = parseInt(check.stdout.trim());
    console.log(`  File size on disk: ${(fileSize / 1024 / 1024).toFixed(1)}MB`);
    if (fileSize === data.length) {
      console.log(`  ✓ Size matches\n`);
    } else {
      console.log(`  ✗ Size mismatch: expected ${data.length}, got ${fileSize}\n`);
    }

    // ── Download via signed URL ────────────────────────────────
    console.log("Getting signed download URL...");
    const downloadUrl = await sb.downloadUrl("/workspace/large.bin");
    console.log(`  URL: ${downloadUrl.slice(0, 80)}...\n`);

    console.log(`Downloading ${sizeMB}MB...`);
    const t1 = Date.now();
    const downloadResp = await fetch(downloadUrl);
    const downloaded = Buffer.from(await downloadResp.arrayBuffer());
    const downloadMs = Date.now() - t1;

    if (downloadResp.ok) {
      const mbps = (sizeMB / (downloadMs / 1000)).toFixed(1);
      console.log(`  ✓ Download: ${downloadMs}ms (${mbps} MB/s)\n`);
    } else {
      console.log(`  ✗ Download failed: ${downloadResp.status}\n`);
      return;
    }

    // Verify hash matches
    const downloadHash = createHash("sha256").update(downloaded).digest("hex").slice(0, 16);
    if (hash === downloadHash) {
      console.log(`  ✓ Hash matches: ${downloadHash}...\n`);
    } else {
      console.log(`  ✗ Hash mismatch: ${hash}... vs ${downloadHash}...\n`);
    }

    console.log("Done — all checks passed!");

  } finally {
    await sb.kill();
    console.log(`\nSandbox ${sb.sandboxId} destroyed.`);
  }
}

main().catch((err) => {
  console.error("Fatal:", err);
  process.exit(1);
});
